### PR TITLE
задание 11

### PR DESCRIPTION
**Pull request by Anastasia-Papchenko**

### CODE REVIEW

**Comment by Anastasia-Papchenko** [comment]
на одном и двух процессах нормально работает, а вот дальше появляются ошибки. Что не так? 
Вот вывод на 4х и 8ю процессах: 

mpirun -np 4 ./Ab --oversubscribe
Vector b:
3.00 6.00 7.00 5.00 3.00 5.00 6.00 2.00
Matrix A:
9.00 1.00 2.00 7.00 0.00 9.00 3.00 6.00
0.00 6.00 2.00 6.00 1.00 8.00 7.00 9.00
3.00 6.00 7.00 5.00 3.00 5.00 6.00 2.00
9.00 1.00 2.00 7.00 0.00 9.00 3.00 6.00
3.00 6.00 7.00 5.00 3.00 5.00 6.00 2.00
9.00 1.00 2.00 7.00 0.00 9.00 3.00 6.00
3.00 6.00 7.00 5.00 3.00 5.00 6.00 2.00
9.00 1.00 2.00 7.00 0.00 9.00 3.00 6.00
Result vector:
82.000000
105.000000
80.000000
145.000000
119.000000
91.000000
82.000000
105.000000
Execution time: 0.000001 seconds

Ab:5701 terminated with signal 11 at PC=7f702ac623fe SP=7ffe973e03e0.  Backtrace:
/lib/x86_64-linux-gnu/libc.so.6(free+0x1e)[0x7f702ac623fe]
/lib/x86_64-linux-gnu/libfabric.so.1(+0x18987)[0x7f702aa24987]
/lib64/ld-linux-x86-64.so.2(+0x624e)[0x7f702d3f824e]
/lib/x86_64-linux-gnu/libc.so.6(+0x45495)[0x7f702ac02495]
/lib/x86_64-linux-gnu/libc.so.6(on_exit+0x0)[0x7f702ac02610]
/lib/x86_64-linux-gnu/libc.so.6(+0x29d97)[0x7f702abe6d97]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80)[0x7f702abe6e40]
./Ab(+0x1305)[0x5616201cb305]


mpirun -np 8 ./Ab --oversubscribe
Vector b:
3.00 6.00 7.00 5.00 3.00 5.00 6.00 2.00
Matrix A:
9.00 1.00 2.00 7.00 0.00 9.00 3.00 6.00
3.00 6.00 7.00 5.00 3.00 5.00 6.00 2.00
3.00 6.00 7.00 5.00 3.00 5.00 6.00 2.00
3.00 6.00 7.00 5.00 3.00 5.00 6.00 2.00
3.00 6.00 7.00 5.00 3.00 5.00 6.00 2.00
3.00 6.00 7.00 5.00 3.00 5.00 6.00 2.00
3.00 6.00 7.00 5.00 3.00 5.00 6.00 2.00
3.00 6.00 7.00 5.00 3.00 5.00 6.00 2.00
Result vector:
82.000000
105.000000
119.000000
91.000000
119.000000
91.000000
119.000000
91.000000
Execution time: 0.000001 seconds
free(): invalid next size (fast)

Ab:5736 terminated with signal 6 at PC=7f9069cda9fc SP=7ffc4bfe7430.  Backtrace:
/lib/x86_64-linux-gnu/libc.so.6(pthread_kill+0x12c)[0x7f9069cda9fc]
/lib/x86_64-linux-gnu/libc.so.6(raise+0x16)[0x7f9069c86476]
/lib/x86_64-linux-gnu/libc.so.6(abort+0xd3)[0x7f9069c6c7f3]
/lib/x86_64-linux-gnu/libc.so.6(+0x89676)[0x7f9069ccd676]
/lib/x86_64-linux-gnu/libc.so.6(+0xa0cfc)[0x7f9069ce4cfc]
/lib/x86_64-linux-gnu/libc.so.6(+0xa2a9d)[0x7f9069ce6a9d]
/lib/x86_64-linux-gnu/libc.so.6(free+0x73)[0x7f9069ce9453]
/lib/x86_64-linux-gnu/libmpich.so.12(+0x1d8dfa)[0x7f906a045dfa]
/lib/x86_64-linux-gnu/libmpich.so.12(+0x21408e)[0x7f906a08108e]
/lib/x86_64-linux-gnu/libmpich.so.12(+0x2143a5)[0x7f906a0813a5]
/lib/x86_64-linux-gnu/libmpich.so.12(+0x215d4b3)[0x7f906bfca4b3]
/lib/x86_64-linux-gnu/libmpich.so.12(MPI_Win_free+0x5b5)[0x7f9069fc6955]
./Ab(+0x1cf5)[0x5593cf452cf5]
/lib/x86_64-linux-gnu/libc.so.6(+0x29d90)[0x7f9069c6dd90]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80)[0x7f9069c6de40]
./Ab(+0x1305)[0x5593cf452305]
double free or corruption (out)

Ab:5736 terminated with signal 6 at PC=7f9069cda9fc SP=7ffc4bfe6800.  Backtrace:
/lib/x86_64-linux-gnu/libc.so.6(pthread_kill+0x12c)[0x7f9069cda9fc]
/lib/x86_64-linux-gnu/libc.so.6(raise+0x16)[0x7f9069c86476]
/lib/x86_64-linux-gnu/libc.so.6(abort+0xd3)[0x7f9069c6c7f3]
/lib/x86_64-linux-gnu/libc.so.6(+0x89676)[0x7f9069ccd676]
/lib/x86_64-linux-gnu/libc.so.6(+0xa0cfc)[0x7f9069ce4cfc]
/lib/x86_64-linux-gnu/libc.so.6(+0xa2e70)[0x7f9069ce6e70]
/lib/x86_64-linux-gnu/libc.so.6(free+0x73)[0x7f9069ce9453]
/lib/x86_64-linux-gnu/libfabric.so.1(+0x1919d)[0x7f9069aac19d]
/lib/x86_64-linux-gnu/libfabric.so.1(+0x188a0)[0x7f9069aab8a0]
/lib64/ld-linux-x86-64.so.2(+0x624e)[0x7f906c47f24e]
/lib/x86_64-linux-gnu/libc.so.6(+0x45495)[0x7f9069c89495]
/lib/x86_64-linux-gnu/libc.so.6(on_exit+0x0)[0x7f9069c89610]
/lib/x86_64-linux-gnu/libinfinipath.so.4(+0x42a7)[0x7f90697b62a7]
/lib/x86_64-linux-gnu/libc.so.6(+0x42520)[0x7f9069c86520]
/lib/x86_64-linux-gnu/libc.so.6(pthread_kill+0x12c)[0x7f9069cda9fc]
/lib/x86_64-linux-gnu/libc.so.6(raise+0x16)[0x7f9069c86476]
/lib/x86_64-linux-gnu/libc.so.6(abort+0xd3)[0x7f9069c6c7f3]
/lib/x86_64-linux-gnu/libc.so.6(+0x89676)[0x7f9069ccd676]
/lib/x86_64-linux-gnu/libc.so.6(+0xa0cfc)[0x7f9069ce4cfc]
/lib/x86_64-linux-gnu/libc.so.6(+0xa2a9d)[0x7f9069ce6a9d]
/lib/x86_64-linux-gnu/libc.so.6(free+0x73)[0x7f9069ce9453]
/lib/x86_64-linux-gnu/libmpich.so.12(+0x1d8dfa)[0x7f906a045dfa]
/lib/x86_64-linux-gnu/libmpich.so.12(+0x21408e)[0x7f906a08108e]
/lib/x86_64-linux-gnu/libmpich.so.12(+0x2143a5)[0x7f906a0813a5]
/lib/x86_64-linux-gnu/libmpich.so.12(+0x215d4b3)[0x7f906bfca4b3]
/lib/x86_64-linux-gnu/libmpich.so.12(MPI_Win_free+0x5b5)[0x7f9069fc6955]
./Ab(+0x1cf5)[0x5593cf452cf5]
/lib/x86_64-linux-gnu/libc.so.6(+0x29d90)[0x7f9069c6dd90]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x80)[0x7f9069c6de40]
./Ab(+0x1305)[0x5593cf452305]

===================================================================================
=   BAD TERMINATION OF ONE OF YOUR APPLICATION PROCESSES
=   PID 5737 RUNNING AT LAPTOP-S30HMKRR
=   EXIT CODE: 9
=   CLEANING UP REMAINING PROCESSES
=   YOU CAN IGNORE THE BELOW CLEANUP MESSAGES
===================================================================================
YOUR APPLICATION TERMINATED WITH THE EXIT STRING: Killed (signal 9)
This typically refers to a problem with your application.
Please see the FAQ page for debugging suggestions

File: Ab.c [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+    for (int i = 0; i < local_rows; i++) {
+        for (int j = 0; j < local_cols; j++) {
+            local_result[i] += local_matrix[i * local_cols + j] * local_vector[j];
+        }
+    }
+
+    double end_time = get_time();
+
+    MPI_Gather(local_result, local_rows, MPI_DOUBLE, result, local_rows, MPI_DOUBLE, 0, MPI_COMM_WORLD);
```
Comment (dimlichmanov) [inline Ab.c]: Проблема сейчас в  этом Gather. Дело не столько в нём, сколько в том, что при числе процессов, начиная с 4, у вас перестаёт быть ленточное разбиение (когда каждый процесс отвечал за свой кусок вектора result), и разбиение становится блочным, и из-за этого за каждый кусок вектора result отвечает уже больше одного процесса
Сам Gather вы делаете правильно, но кладёте всё это в вектор result, который может к себе поместить только local_rows * dims[0] элементов, а не local_rows * size.

**Review by dimlichmanov — CHANGES_REQUESTED** [review]
.

**Comment by dimlichmanov** [comment]
про polus гляну в течение дня

**Pull request closed without merging**